tensorboard/plugins/beholder/tf_beholder_dashboard/tf-beholder-video.html

```diff
@@ -37,50 +37,81 @@
-    const PING_ROUTE = tf_backend.getRouter().pluginRoute('beholder', '/ping')
-
-    var connectionFailed = false
+    const BLANK = 'data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=';
+    const FEED_ROUTE = tf_backend.getRouter().pluginRoute('beholder', '/beholder-frame');
+    const PING_ROUTE = tf_backend.getRouter().pluginRoute('beholder', '/ping');
 
     Polymer({
       is: "tf-beholder-video",
       properties: {
-        _requestManager: {
-          type: Object,
-          value: () => new tf_backend.RequestManager(),
+        pingSleep: {
+          type: Number,
+          value: 1000,
+        },
+
+        xhrTimeout: {
+          type: Number,
+          value: 2500,
         },
 
         _imageURL: {
           type: String,
-          value: FEED_ROUTE
-        }
+          value: BLANK,
+        },
+
+        _xhr: Object,
+        _timer: Object,
+        _isDead: Boolean,
+      },
+
+      attached() {
+        this.set('_imageURL', FEED_ROUTE);
+        this._ping();
       },
 
-      ready() {
-        // Periodically check the connection, restart if needed.
-        window.setInterval(function(){
-          this._requestManager.request(PING_ROUTE).then(
-              function(response){
-                if (connectionFailed) {
-                  this.restartVideo()
-                }
-                connectionFailed = false
-              }.bind(this),
-              function(error){
-                connectionFailed = true
-              }.bind(this))
-        }.bind(this), 1000)
+      detached() {
+        this._clear();
       },
 
-      restartVideo() {
-        var newImage = document.createElement('img')
-        newImage.src = FEED_ROUTE + '?t=' + new Date().getTime()
-        newImage.id = 'video'
-        newImage.className = 'style-scope beholder-video'
-        this.$.container.appendChild(newImage)
+      _ping() {
+        this._clear();
+        this._xhr = new XMLHttpRequest();
+        this._xhr.open('GET', PING_ROUTE, true);
+        this._xhr.timeout = this.xhrTimeout;
+        this._xhr.onload = this._onPingLoad.bind(this);
+        this._xhr.onerror = this._onPing.bind(this, false, this.pingSleep);
+        this._xhr.ontimeout = this._onPing.bind(this, false, 1);
+        this._xhr.send(null);
+      },
+
+      _onPingLoad() {
+        if (this._xhr.status == 200) {
+          const response = JSON.parse(this._xhr.responseText);
+          this._onPing(response['status'] == 'alive', this.pingSleep);
+          return;
+        } 
+        this._onPing(false, this.pingSleep);
+      },
+
+      _onPing(isAlive, retryMs) {
+        if (isAlive && this._isDead) {
+          this.set('_imageURL', FEED_ROUTE + '?t=' + new Date().getTime());
+        }
+        this._isDead = !isAlive;
+        this._timer = window.setTimeout(this._ping.bind(this), retryMs);
+      },
 
-        this.$.container.removeChild(this.$.video)
-        this.$.video = newImage
+      _clear() {
```

Comment:
Detaching this element from the DOM doesn't appear to be enough to get it garbage-collected, and even after it's detached, it continues to consume frames from the backend server, which consumes unnecessary resources both client-side (visible in chrome://net-internals) as well as one the server.

The logic that clears the video resources should therefore also set _imageURL back to the BLANK src value, which should at least more reliably interrupt the stream and free those resources.

Reply:
Now that's a fantastic catch. Setting it to BLANK on detach caused the framed PNG to stop downloading, thereby saving 250kBps. Confirmed with chrome://net-internals/#timeline. I'm more concerned with traffic than DOM memory right now.